### PR TITLE
Allow API to create file on empty repo

### LINKED ISSUE #10993: [Feature] [API] Initialize Empty repo if repoCreateFile is used
as title

related bug for this api: #10982

### PR DESCRIPTION
**Pull request by zeripath**
This PR adds the necessary work to make it possible to create files on empty
repos using the API.

Fix #10993

Signed-off-by: Andrew Thornton <art27@cantab.net>

### CODE REVIEW

File: services/repository/files/temp_repo.go [hunk 1]
```diff
@@ -235,11 +248,23 @@ func (t *TemporaryUploadRepository) CommitTreeWithDate(author, committer *user_m
...
+	}
 
 	// Determine if we should sign
 	if git.CheckGitVersionAtLeast("1.7.9") == nil {
-		sign, keyID, signer, _ := asymkey_service.SignCRUDAction(t.ctx, t.repo.RepoPath(), author, t.basePath, "HEAD")
+		var sign bool
+		var keyID string
+		var signer *git.Signature
+		if parent != "" {
+			sign, keyID, signer, _ = asymkey_service.SignCRUDAction(t.ctx, t.repo.RepoPath(), author, t.basePath, "HEAD")
```
Comment (zeripath) [inline services/repository/files/temp_repo.go]:
```suggestion
			sign, keyID, signer, _ = asymkey_service.SignCRUDAction(t.ctx, t.repo.RepoPath(), author, t.basePath, parent)
```

File: services/repository/files/temp_repo.go [hunk 2]
```diff
@@ -77,6 +77,19 @@ func (t *TemporaryUploadRepository) Clone(branch string) error {
 	return nil
 }
 
+// Open the repository
```
Comment (zeripath) [inline services/repository/files/temp_repo.go]:
```suggestion
// Init the repository
```

**Comment by lunny** [comment]
I think we need a testcase for empty repository.

**Comment by zeripath** [comment]
done

File: services/repository/files/update.go [hunk 3]
```diff
@@ -191,118 +191,130 @@ func CreateOrUpdateRepoFile(ctx context.Context, repo *repo_model.Repository, do
...
-					CurrentSHA: fromEntry.ID.String(),
-				}
+
+		// Assigned LastCommitID in opts if it hasn't been set
+		if opts.LastCommitID == "" {
+			opts.LastCommitID = commit.ID.String()
+		} else {
+			lastCommitID, err := t.gitRepo.ConvertToSHA1(opts.LastCommitID)
+			if err != nil {
+				return nil, fmt.Errorf("DeleteRepoFile: Invalid last commit ID: %v", err)
```
Comment (Gusted) [inline services/repository/files/update.go]:
```suggestion
				return nil, fmt.Errorf("ConvertToSHA1: Invalid last commit ID: %v", err)
```

**Review by Gusted — APPROVED** [review]
Otherwise seems to LGTM.

File: services/repository/files/patch.go [hunk 4]
```diff
@@ -164,9 +164,9 @@ func ApplyDiffPatch(ctx context.Context, repo *repo_model.Repository, doer *user
 	// Now commit the tree
 	var commitHash string
 	if opts.Dates != nil {
-		commitHash, err = t.CommitTreeWithDate(author, committer, treeHash, message, opts.Signoff, opts.Dates.Author, opts.Dates.Committer)
+		commitHash, err = t.CommitTreeWithDate("HEAD", author, committer, treeHash, message, opts.Signoff, opts.Dates.Author, opts.Dates.Committer)
```
Comment (6543) [inline services/repository/files/patch.go:167]:
```suggestion
		commitHash, err = t.CommitTreeWithDate(repo.DefaultBranch, author, committer, treeHash, message, opts.Signoff, opts.Dates.Author, opts.Dates.Committer)
```
?

Comment (Gusted) [reply services/repository/files/patch.go:167]: The "current" behavior is using "HEAD", https://github.com/go-gitea/gitea/pull/19224/files#diff-b0213d5f52f1ddd2d9701c3f073f147a0a18f76d8975605dc5037a377a25f24aL238 

It uploads to a temporary repository, given it clones to a new repo and the existing repo can have any kind of default branch name the "HEAD" seems to be most correct here(given we just clone, so the HEAD should be set correct).

Comment (6543) [reply services/repository/files/patch.go:167]: noticed while reading code ...

**Review by 6543 — APPROVED** [review]

**Pull request merged**